prov/rxm/src/rxm_conn.c

```diff
@@ -117,34 +117,94 @@ rxm_conn_send_queue_close(struct rxm_conn *rxm_conn)
 
 static void rxm_conn_close(struct util_cmap_handle *handle)
 {
-	struct rxm_conn *rxm_conn = container_of(handle, struct rxm_conn, handle);
+	struct rxm_conn *rxm_conn =
+		container_of(handle, struct rxm_conn, handle);
 
 	if (!rxm_conn->msg_ep)
 		return;
+
+	if (handle->cmap->attr.serial_access) {
+		if (fi_close(&rxm_conn->msg_ep->fid)) {
+			FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+				"Unable to close msg_ep\n");
+		} else {
+			FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
+			       "Closed msg_ep\n");
+		}
+	} else {
+		rxm_conn->saved_msg_ep = rxm_conn->msg_ep;
+		FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
+		       "Saved MSG EP fid for further deletion in main thread\n");
+	}
+	rxm_conn->msg_ep = NULL;
+}
+
+static void rxm_conn_save(struct util_cmap_handle *handle)
+{
+	struct rxm_conn *rxm_conn =
+		container_of(handle, struct rxm_conn, handle);
+
+	if (!rxm_conn->msg_ep)
+		return;
+
 	rxm_conn->saved_msg_ep = rxm_conn->msg_ep;
 	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
-	       "Saved MSG EP fid for further deletion in main thread\n");
+	       "Saved MSG EP fid for further deletion\n");
 	rxm_conn->msg_ep = NULL;
 }
 
+static void rxm_conn_close_saved(struct util_cmap_handle *handle)
+{
+	struct rxm_conn *rxm_conn =
+		container_of(handle, struct rxm_conn, handle);
+
+	if (!rxm_conn->saved_msg_ep)
+		return;
+
+	if (handle->cmap->attr.serial_access) {
+		if (fi_close(&rxm_conn->saved_msg_ep->fid)) {
+			FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+				"Unable to close saved msg_ep\n");
+		} else {
+			FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
+			       "Closed saved msg_ep\n");
+		}
+		rxm_conn->saved_msg_ep = NULL;
+	} else {
+		rxm_conn->saved_msg_ep = rxm_conn->msg_ep;
```

Comment:
Wouldn't this overwrite rxm_conn->saved_msg_ep? I think we don't need the else case here.

Reply:
Good catch! Thanks!